First pass on JOSS paper (#72)
invenia/JLSO.jl: JOSS/paper.md

---
title: 'JLSO: Storing reproducible serialized objects in Julia'
tags:
  - julialang
  - data storage
  - reproducibility
authors:
  - name: Rory Finnegan
    orcid: 0000-0002-4603-077X
    affiliation: 1
  - name: Lyndon White

>     orcid: 0000-0003-1386-1646

Reviewer: That is indeed my DOI number

Author: Yeah, I looked at the tensorflow paper.

    affiliation: 2
affiliations:

>  - name: Research Engineer, Invenia Technical Computing

Reviewer: Position names don't generally go in affliations
```suggestion
 - name: Invenia Technical Computing
```

   index: 1

>  - name: Research Engineer, Invenia Labs

Reviewer: ```suggestion
 - name: Invenia Labs
```

   index: 2
date: 24 August 2022
bibliography: paper.bib
---

# Summary

> As scientific computing software grows increasingly complex, the need to efficiently and reliably store sophisticated program objects has become a growing need. The expanding list of file formats for serializing objects is evidence of this problem. Unfortunately, these file formats typically come with usability versus reliability tradeoffs.

Reviewer: ```suggestion
As scientific computing software grows increasingly complex, the need to efficiently and reliably store sophisticated program objects likewise increases.
The expanding list of file formats for serializing objects is evidence of this problem.
Unfortunately, these file formats typically come with usability versus reliability tradeoffs.
```

First sentence seem repeditive.

Also, i think 1 sentence per line is best here (like in docs) so that diffs stay sensible.
I am not going to go an suggest changes everywhere for that.

Author: Alright, I think all the sentences are on separate lines now. I've also rewritten that first sentence to make it less repetitive.

> Choosing serialization formats such as CSV, JSON, BSON, and HDF5 are prudent choices for long term storage because they are application and language agnostic. Projects that evolve beyond an initial language or library decision can still load old experimental data.

Reviewer: ```suggestion
Generally, choising application and language agnostic serialization formats, such as CSV, JSON, BSON, and HDF5, is prudent for long term storage.
Projects that evolve beyond an initial language or library decision can still load old experimental data.
```

> These formats often support a limited set of types, requiring applications to define and maintain serialization methods for their custom objects.

Reviewer: ```suggestion
These formats only ever support a limited set of types in the standard.
This requires applications to define and maintain their own serialization methods for objects outside that standard set.
```

> Using formats such as JLD (Julia Data), Python pickles or MAT files work best for convenient saving of arbitrary objects. Unfortunately, these formats are often highly coupled to the software dependencies (e.g., language version, software packages), which makes restoring the data more challenging as time passes and the software evolves.

Reviewer: Probably we should mention specifically in this section that JLD and MAT are extensions of HDF5, and that while this has some advantages of HDF5, it falls back in other cases to being no better than pickle or the julia serializer.
We should mention fully bespoke things like Pickle and the julia serialiser first.

Reviewer: We should make it clear that this JLSO is firmly inside the second category, and is vulnerable to all the same problems.
but that it seaks to amiliorate them via keeping enough metadata around to recover.

Author: Hmmm, should we maybe have 3 categories then?

1. General/agnostic file formats (e.g., CSV, BSON, HDF5)
2. Language specific serializers like pickles or the julia serializer
3. File formats that merge the two. Arguably JLSO fits into this category, but in a different way.

Didn't MAT files only extend the HDF spec in 7.3 though?

Reviewer: Yes, but MatLab  7.3 came out 14 years ago now.

Reviewer: I would say 1 and 2 exist.

and that .JLD, .MAT, and BSON.jl are special cases of 2
but that for the right inputs may be partially or totally readable with tooling for 1.

and JLSO is a related but different case to that.

Author: > MatLab 7.3 came out 14 years ago now.

Lol, good point.

> JLSO is a related but different case to that.

Okay, I've opted to remove the reference to JLD and MAT for now. I think this simplifies the discussion, but I can add it back if you think it's worth distinguishing JLSO from those specific formats.

```julia
julia> using Pkg, JLSO

julia> jlso = read("../test/specimens/v3_julia_serialize_none.jlso", JLSOFile)
JLSOFile([ZonedDateTime, DataFrame, Vector, DateTime, String, Matrix, Distribution]; version="3.0.0", julia="1.0.3", format=:julia_serialize, compression=:none, image="")

# Examples of different julia objects that can be serialized
julia> names(jlso)
7-element Array{Symbol,1}:
 :ZonedDateTime
 :DataFrame
 :Vector
 :DateTime
 :String
 :Matrix
 :Distribution

julia> jlso[:Matrix]
3×3 Array{Float64,2}:
 0.400348   0.892196   0.848164
 0.0183529  0.755449   0.397538
 0.870458   0.0441878  0.170899

julia> jlso[:String]
"Hello World!"

julia> jlso[:DateTime]
2018-01-28T00:00:00

julia> jlso[:Distribution];
[warn | JLSO]: KeyError: key Distributions [31c24e10-a181-5473-b8eb-7969acd0382f] not found

julia> using Distributions
ERROR: ArgumentError: Package Distributions not found in current path:
- Run `import Pkg; Pkg.add("Distributions")` to install the Distributions package.

Stacktrace:
 [1] require(::Module, ::Symbol) at ./loading.jl:893
```

> As you can see, we do not have Distributions installed and we also don't know which version of Distributions.jl we need to successfully deserialize that object. As of version 1.2.0, users can resolve these kinds of concerns by rebuilding the package environment used to save a JLSO file directly from the REPL.

Reviewer: ```suggestion
As you can see, we do not have Distributions installed and we also don't know which version of Distributions.jl we need to successfully deserialize that object.
Users can resolve these kinds of concerns by rebuilding the package environment used to save a JLSO file directly from the REPL.
```
We don't need to mention features not present in older versions, noone is going to use older versions.

```julia
julia> using Pkg

julia> Pkg.activate(jlso, "./")
 Activating environment at `~/repos/invenia/JLSO.jl/JOSS/Project.toml`

# Now we can process data within this file using the environment used to save it.
julia> using Distributions

julia> jlso[:Distribution]
Normal{Float64}(μ=50.2, σ=4.3)
```

> While the metadata in JLSO files caters towards Julia users, the format itself is mostly language agnostic. A variety of internal object serialization formats can be used, and the metadata itself is saved as a BSON documented. The Julia language provides first-class support for package environments via Project.toml and Manifest.toml files, which provided an intuitive platform for building our prototype.

Reviewer: I would restructure this paragraph.
Posibly into two?

Somehere in here should be implementation stuff.
"JLSO achives this via working with a language dependent serializer (julia serialization stdlib, or the extended BSON.jl format), and pairing that with metadata stored ina language independent BSON top level ...)
and that the concept of this format is is not julia specific, but JLSO.jl is.

We also want to avoid being so unclear as "the format itself is mostly language agnostic"
